internal/host/plugin/repository_host_set.go

```diff
@@ -152,6 +156,364 @@ func (r *Repository) CreateSet(ctx context.Context, scopeId string, s *HostSet,
+							[]string{"version"},
+							[]string{},
+							db.NewOplogMsg(&hsOplogMsg),
+							db.WithVersion(&version),
+						)
+						if err != nil {
+							return errors.Wrap(ctx, err, op)
+						}
+
+						if numUpdated != 1 {
+							return errors.New(ctx, errors.MultipleRecords, op, fmt.Sprintf("expected 1 set to be updated, got %d", numUpdated))
+						}
+
+						recordUpdated = true
+						msgs = append(msgs, &hsOplogMsg)
+					}
+
+					msgs = append(msgs, &peDeleteOplogMsg)
+				}
+
+			case endpointOpUpdate:
+				// Delete all old endpoint entries.
+				var peDeleteOplogMsg oplog.Message
+				pep := &host.PreferredEndpoint{
+					PreferredEndpoint: &store.PreferredEndpoint{
+						HostSetId: newSet.PublicId,
+						Priority:  1,
+					},
+				}
+				_, err := w.Delete(ctx, pep, db.WithWhere("host_set_id = ?", newSet.PublicId), db.NewOplogMsg(&peDeleteOplogMsg))
+				if err != nil {
+					return errors.Wrap(ctx, err, op)
+				}
+
+				// Only append the oplog message if an operation was actually
+				// performed.
+				if peDeleteOplogMsg.Message != nil {
+					msgs = append(msgs, &peDeleteOplogMsg)
+				}
+
+				// Create the new entries.
+				peCreateOplogMsgs := make([]*oplog.Message, 0, len(preferredEndpoints))
+				if err := w.CreateItems(ctx, preferredEndpoints, db.NewOplogMsgs(&peCreateOplogMsgs)); err != nil {
+					return err
+				}
+
+				if !recordUpdated {
+					// we only updated secrets, so we need to increment the
+					// version of the host catalog manually.
+					returnedSet = newSet.clone()
+					returnedSet.Version = uint32(version) + 1
+					var hsOplogMsg oplog.Message
+					numUpdated, err := w.Update(
+						ctx,
+						returnedSet,
+						[]string{"version"},
+						[]string{},
+						db.NewOplogMsg(&hsOplogMsg),
+						db.WithVersion(&version),
+					)
+					if err != nil {
+						return errors.Wrap(ctx, err, op)
+					}
+
+					if numUpdated != 1 {
+						return errors.New(ctx, errors.MultipleRecords, op, fmt.Sprintf("expected 1 set to be updated, got %d", numUpdated))
+					}
+
+					recordUpdated = true
+					msgs = append(msgs, &hsOplogMsg)
+				}
+
+				msgs = append(msgs, peCreateOplogMsgs...)
+			}
+
+			if !pluginCalledSuccessfully {
+				_, err = plgClient.OnUpdateSet(ctx, &plgpb.OnUpdateSetRequest{
+					CurrentSet: currentPlgSet,
+					NewSet:     newPlgSet,
+					Catalog:    plgHc,
+					Persisted:  persisted,
+				})
+				if err != nil {
+					if status.Code(err) != codes.Unimplemented {
+						return errors.Wrap(ctx, err, op)
+					}
+				}
+			}
+
+			if len(msgs) != 0 {
+				metadata := s.oplog(oplog.OpType_OP_TYPE_UPDATE)
+				if err := w.WriteOplogEntryWith(ctx, oplogWrapper, ticket, metadata, msgs); err != nil {
+					return errors.Wrap(ctx, err, op, errors.WithMsg("unable to write oplog"))
+				}
+			}
+
+			if returnedSet == nil {
+				// Nothing updated, so returned set will not have been
+				// cloned.  Clone it now so that we can fetch hosts from it.
+				returnedSet = newSet.clone()
```

Comment:
Instead of doing this repeatedly throughout the transaction, just do it once at the very beginning unconditionally.

Reply:
Done!